main/refer.c

```diff
@@ -0,0 +1,543 @@
+		return NULL;
+	}
+
+	if (ast_string_field_init(refer, 128)) {
+		ao2_ref(refer, -1);
+		return NULL;
+	}
+
+	refer->vars = ao2_container_alloc_list(AO2_ALLOC_OPT_LOCK_MUTEX, 0,
+		NULL, refer_data_cmp_fn);
+	if (!refer->vars) {
+		ao2_ref(refer, -1);
+		return NULL;
+	}
+	refer->to_self = 0;
+
+	return refer;
+}
+
+struct ast_refer *ast_refer_ref(struct ast_refer *refer)
+{
+	ao2_ref(refer, 1);
+	return refer;
+}
+
+struct ast_refer *ast_refer_destroy(struct ast_refer *refer)
+{
+	ao2_ref(refer, -1);
+	return NULL;
+}
+
+int ast_refer_set_to(struct ast_refer *refer, const char *fmt, ...)
+{
+	va_list ap;
+
+	va_start(ap, fmt);
+	ast_string_field_build_va(refer, to, fmt, ap);
+	va_end(ap);
+
+	return 0;
+}
+
+int ast_refer_set_from(struct ast_refer *refer, const char *fmt, ...)
+{
+	va_list ap;
+
+	va_start(ap, fmt);
+	ast_string_field_build_va(refer, from, fmt, ap);
+	va_end(ap);
+
+	return 0;
+}
+
+int ast_refer_set_refer_to(struct ast_refer *refer, const char *fmt, ...)
+{
+	va_list ap;
+
+	va_start(ap, fmt);
+	ast_string_field_build_va(refer, refer_to, fmt, ap);
+	va_end(ap);
+
+	return 0;
+}
+
+int ast_refer_set_to_self(struct ast_refer *refer, int val)
+{
+	refer->to_self = val;
+	return 0;
+}
+
+int ast_refer_set_tech(struct ast_refer *refer, const char *fmt, ...)
+{
+	va_list ap;
+
+	va_start(ap, fmt);
+	ast_string_field_build_va(refer, tech, fmt, ap);
+	va_end(ap);
+
+	return 0;
+}
+
+int ast_refer_set_endpoint(struct ast_refer *refer, const char *fmt, ...)
+{
+	va_list ap;
+
+	va_start(ap, fmt);
+	ast_string_field_build_va(refer, endpoint, fmt, ap);
+	va_end(ap);
+
+	return 0;
+}
+
+const char *ast_refer_get_refer_to(const struct ast_refer *refer)
+{
+	return refer->refer_to;
+}
+
+const char *ast_refer_get_from(const struct ast_refer *refer)
+{
+	return refer->from;
+}
+
+const char *ast_refer_get_to(const struct ast_refer *refer)
+{
+	return refer->to;
+}
+
+int ast_refer_get_to_self(const struct ast_refer *refer)
+{
+	return refer->to_self;
+}
+
+const char *ast_refer_get_tech(const struct ast_refer *refer)
+{
+	return refer->tech;
+}
+
+const char *ast_refer_get_endpoint(const struct ast_refer *refer)
+{
+	return refer->endpoint;
+}
+
+static struct refer_data *refer_data_new(const char *name)
+{
+	struct refer_data *data;
+	int name_len = strlen(name) + 1;
+
+	if ((data = ao2_alloc_options(name_len + sizeof(*data), refer_data_destructor, AO2_ALLOC_OPT_LOCK_NOLOCK))) {
+		data->name = data->stuff;
+		strcpy(data->name, name);
+	}
+
+	return data;
+}
+
+static struct refer_data *refer_data_find(struct ao2_container *vars, const char *name)
+{
+	struct refer_data tmp = {
+		.name = (char *)name,
+	};
+	return ao2_find(vars, &tmp, OBJ_SEARCH_KEY);
+}
+
+char *ast_refer_get_var_and_unlink(struct ast_refer *refer, const char *name)
+{
+	struct refer_data *data;
+	char *val = NULL;
+	struct refer_data tmp = {
+		.name = (char*)name,
+	};
+
+	if (!(data = ao2_find(refer->vars, &tmp, OBJ_SEARCH_KEY | OBJ_UNLINK))) {
```

Comment:
This is incorrect. You don't need a temporary structure, you can pass name to ao2_find.

Reply:
Thank you! Fixed.